components/automate-gateway/gateway/middleware/auth_interceptor.go

```diff
@@ -98,6 +98,60 @@ type authInterceptor struct {
 	authz GRPCAuthorizationHandler
 }
 
+func (a *authInterceptor) combinedAuth(ctxIn context.Context, req interface{}) (context.Context, error) {
+	// extract request-scoped logger
+	log := ctxlogrus.Extract(ctxIn)
+
+	// this context is only used for authenticating the request: we need
+	// headers!
+	// grpc-gateway translates these into metadata, and we'll forward that to
+	// authn-service below. we don't want that metadata on _every_ request,
+	// though, so this context is separated out
+	authCtx := ctxIn
+
+	// transfer incoming metadata to outgoing metadata
+	// if !ok, this will default to the ctx set above, and most likely be
+	// rejected by authn-service. However, we don't want to duplicate the
+	// "when is a request potentially good for authentication" logic here,
+	// so this will just go to authn-service as-is.
+	var md metadata.MD
+	var ok bool
+	if md, ok = metadata.FromIncomingContext(authCtx); ok {
+		authCtx = metadata.NewOutgoingContext(authCtx, md)
+	}
+
+	var subs []string
+	var authResponse *authn.AuthenticateResponse
+	var err error
+
+	certSubject, ok := headerAuthValidForClientAndPeer(ctxIn)
+	if ok && !fromGateway(certSubject) {
+		subs = []string{certSubject}
+		log.Debugf("using client cert to authenticate request: %q", certSubject)
+	} else { // ordinary authn by calling authn-service
+		authResponse, err = a.authn.Authenticate(authCtx, &authn.AuthenticateRequest{})
+		if err != nil {
+			log.Debugf("error authenticating request: %s", err)
+			return nil, err
+		}
+		subs = append(authResponse.Teams, authResponse.Subject)
+	}
+
+	projects := auth_context.ProjectsFromMetadata(md)
+
+	ctxOut, err := a.authz.Handle(authCtx, subs, projects, req)
```

Comment:
Would appreciate some double-checking that I have the right contexts going in and coming out of the various places

Reply:
That looks right to me, but maybe worth adding some testing to that effect. It'd also be nice to see this in action with some e2e tests / somewhere we could see the context we are expecting show up. Hard to validate this behavior without an upstream.